### PR TITLE
Add tests for ouimeaux_device.service.api.Action class

### PR DESCRIPTION
**Pull request by sullivanmj**
## Description:
With the eventual changes to pyWeMo of deprecating python 2 support and adding async functionality, I started adding some unit tests, as well as integrating them with the Travis CI. I had some thoughts & questions along the way. I figured it would be easier to discuss these things if there was a working pull request rather than just opening another issue. Especially since adding the first tests will, set precedent for tests that follow. This is the first of hopefully many PRs to add more tests so that we can make sure we're not breaking things along the way.

Anyway, in no particular order some possible points of discussion:

- It appears there was some previous configuration for running pytest with Travis CI, so I was able to reuse that without making too significant of changes to the Travis CI & Tox configuration.
- It seems there are quite a few ways to organize test packages/modules with respect to the code that they test. It was mostly a coinflip for me as to which direction to go, so I'm open to changing that approach.
- I took the liberty of excluding the tests from the existing flake8 call, and adding a separate call to that specific directory with some exclusions for docstrings. I understand if you don't want that somewhat unnecessary complexity in the build.
- For some reason, on my machine, flake8 seemed to be picking up its configuration from tox.ini, where .venv was not excluded - making it take quite a while to execute. This did not appear to be happening on the Travis CI. Perhaps it pulls flake8 config from setup.cfg instead? Anyway, that's why I added that exclusion from the tox.ini - to prevent the pain locally.
- I attempted to squash my commits into some more intelligible/less noisy commits, but now that I see them in the PR form, it might not have worked? Maybe that's just how github works, or maybe I did it wrong, or maybe something else? Either way - I'm not sure if it's a big deal or not.

**Related issue (if applicable):** n/a

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] There is no commented out code in this PR.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -2,7 +2,7 @@
 envlist = lint, pylint
 skip_missing_interpreters = True
 
-[testenv]
+[testenv:pytest]
```
Comment (sqldiablo) [inline tox.ini]: Can we keep this named [testenv]? That will line up with some other changes I was planning to make to our testing/linting setup.

**Review by sqldiablo — CHANGES_REQUESTED** [review]

File: tests/ouimeaux_device/api/unit/test_service.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+from xml.etree import cElementTree as cet
+from xml.etree import ElementTree
+import unittest.mock as mock
+import pytest
+import requests
+
+import pywemo.ouimeaux_device.api.service as svc
+
+HEADERS_KWARG_KEY = 'headers'
```
[the remark this answers is not in the document]
Comment (sqldiablo) [reply tests/ouimeaux_device/api/unit/test_service.py]: Can you run black against this?

black --fast pywemo tests

File: tests/ouimeaux_device/api/unit/test_service.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        try:
+            action()
+        except svc.ActionException:
+            pass
+
+        assert post_mock.call_count == svc.MAX_RETRIES
+
+    def test_call_throws_when_final_retry_fails(self):
+        action = self.get_mock_action()
+        requests.post = mock.Mock(side_effect=requests.exceptions.RequestException)
```
[the remark this answers is not in the document]
Comment (sullivanmj) [reply tests/ouimeaux_device/api/unit/test_service.py]: Ha! Black made this line too long. Go figure.

File: tests/ouimeaux_device/api/unit/test_service.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        try:
+            action()
+        except svc.ActionException:
+            pass
+
+        assert post_mock.call_count == svc.MAX_RETRIES
+
+    def test_call_throws_when_final_retry_fails(self):
+        action = self.get_mock_action()
+        requests.post = mock.Mock(
```
[the remark this answers is not in the document]
Comment (sullivanmj) [reply tests/ouimeaux_device/api/unit/test_service.py:118]: Apparently black/hound are at odds on line length... See [this comment](https://github.com/pavoni/pywemo/pull/153#pullrequestreview-421528798).

File: tox.ini (hunk 1 above)
Comment (sullivanmj) [reply tox.ini]: Sure. That was the only way that I could figure out how to get pytest to run in the Travis CI build with the existing Tox configuration. Could you suggest how to get that working with the original configuration here? I haven't worked with Tox before now.

File: tests/ouimeaux_device/api/unit/test_service.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sqldiablo) [reply tests/ouimeaux_device/api/unit/test_service.py:118]: lol, that's a good time. Go ahead and edit .hound.yml and set enabled to false for now. I'm more concerned with flake8 and the other tests that Travis runs for us.

File: tox.ini (hunk 1 above)
Comment (sqldiablo) [reply tox.ini]: Try adding py35 to the envlist.

Comment (sqldiablo) [reply tox.ini]: Let's try making it look like this:

```ini
[testenv]
basepython = {env:PYTHON3_PATH:python3}
commands =
     pytest {posargs}
     {toxinidir}/script/check_dirty
deps =
     -r{toxinidir}/requirements_test.txt
     pytest-xdist
```

Comment (sqldiablo) [reply tox.ini]: Scratch that, you found the issue.

**Comment by sqldiablo** [comment]
This looks good to me, if it looks good to you at this point.

**Review by sqldiablo — APPROVED** [review]

**Comment by sullivanmj** [comment]
Yeah, as long as you're OK with the direction I went with setting up the tests, I'm happy with it.

**Comment by sqldiablo** [comment]
I don't have much experience with tests in python because I'm mainly just a hobbyist python developer. What you did seems to make sense, so I'm ok with it. I appreciate your help with getting us started with tests.

**Pull request merged**